packages/new-deploy/src/DeployContext.ts

```diff
@@ -0,0 +1,310 @@
+  ): void {
+    const name = this.resolveDomainName(nameOrDomain);
+    if (!this.verification.has(name)) this.verification.set(name, []);
+
+    const net = this._verification.get(name);
+    if (net) net.push(verification);
+  }
+
+  protected async deployCore(domain: config.Domain): Promise<void> {
+    this.addDomain(domain);
+
+    const core = new CoreContracts(this, domain.name);
+    await core.recordStartBlock();
+
+    await Promise.all([
+      core.deployUpgradeBeaconController(),
+      core.deployUpdaterManager(),
+      core.deployXAppConnectionManager(),
+    ]);
+
+    await core.deployHome();
+    await core.deployGovernanceRouter();
+
+    // all contracts deployed
+    const complete = core.complete();
+    this.addCore(domain.name, complete);
+  }
+
+  // post-deployment setup
+  // We choose to enroll all currently known routers and watchers here,
+  // event those that may not be immediately connected, so that IF they are
+  // connected in the future, the routers will be properly configured.
+  // Essentially we run likely future governance actions at deploy time, so
+  // we will not be required to run them later.
+  protected async enrollCores(): Promise<ethers.PopulatedTransaction[]> {
+    const results = await Promise.all(
+      this.networks.map(async (network) => {
+        // the set of domains that are not the new core
+        const remoteDomains = this.data.networks.filter(
+          (net) => net !== network,
+        );
+        const core = this.mustGetCore(network);
+
+        return (
+          await Promise.all([
+            ...remoteDomains.map((remote) =>
+              core.enrollGovernanceRouter(remote),
+            ),
+            ...remoteDomains.map((remote) => core.enrollWatchers(remote)),
+          ])
+        ).flat();
+      }),
+    );
+    return results.flat();
+  }
+
+  protected async deployBridge(name: string): Promise<void> {
+    this.mustGetCore(name).complete(); // assert that the core is totally deployed
+
+    const bridge = new BridgeContracts(this, name);
+    await bridge.recordStartBlock();
+
+    await Promise.all([
+      bridge.deployTokenUpgradeBeacon(),
+      bridge.deployTokenRegistry(),
+      bridge.deployBridgeRouter(),
+      bridge.deployEthHelper(),
+    ]);
+
+    this.addBridge(name, bridge.complete());
+  }
+
+  /// Deploys all configured Cores
+  async ensureCores(): Promise<ethers.PopulatedTransaction[]> {
+    const toDeploy = this.networks.filter((net) => !this.cores[net]);
+
+    await Promise.all(
+      toDeploy.map((net) => this.deployCore(this.mustGetDomainConfig(net))),
+    );
+    return await this.enrollCores();
+  }
+
+  // Deploys all configured connections and enrolls them if possible.
+  // For any connection that cannot be enrolled, outputs the governance
+  // action required to enroll them
+  async ensureConnections(): Promise<ethers.PopulatedTransaction[]> {
+    this.validate();
+
+    const ensure = await this.ensureCores();
```

Comment:
During this call, one of the steps is to `enrollCores` and then `enrollWatchers`. Also inside of this method, in the end, we `enrollRemotes`, where inside we `enrollWatchers` again. Maybe inside of the `enrollWatchers` method (and potentially similar others) we could include a check that some job is done again. It makes sense that some components are reused in different scenarios, but using checks could also save some gas

Reply:
I created an issue https://github.com/nomad-xyz/monorepo/issues/84

Reply:
Cool. I have fixed that for the watchers specifically by adding a filtering step in [2013d53](https://github.com/nomad-xyz/monorepo/pull/15/commits/2013d53206577bfc94f31770845da97eb60be122)

Reply:
most contract deployments will not be redone as this check already exists. same for relinquishing